### PR TITLE
Deterministic backends and asserts system for vfc_ci

### PR DESCRIPTION
**Pull request by PurplePachyderm**
This PR brings two important features for Verificarlo CI : 

- **Support for deterministic backends** (such as VPREC) : if a backend is considered deterministic, a large part of the data processing will be ignored (computation of significant digits, etc ...), and results from this backend will be shown separately in the report.
- **Asserts system** : Probes can now optionally be linked with an "assert". This is corresponds to an accuracy target for the variable, which will be checked in preprocessing. Asserts can be used with both deterministic and non-deterministic backends, and can be of absolute or relative type. For each case, the way to validate the probe is different.

Documentation has also been updated to explain those features more in details.

### CODE REVIEW

File: src/common/vfc_probes.c [hunk 1]
```diff
@@ -209,14 +246,15 @@ int vfc_dump_probes(vfc_probes *probes) {
   }
 
   // First line gives the column names
-  fprintf(fp, "test,variable,value\n");
+  fprintf(fp, "test,variable,value,accuracy_threshold,assert_mode\n");
 
   // Iterate over all table elements
   vfc_probe_node *probe = NULL;
-  for (size_t i = 0; i < probes->map->capacity; i++) {
+  for (int i = 0; i < probes->map->capacity; i++) {
```
Comment (pablooliveira) [inline src/common/vfc_probes.c:253]: Better use `size_t`.

File: src/common/vfc_probes.c [hunk 2]
```diff
@@ -209,14 +246,15 @@ int vfc_dump_probes(vfc_probes *probes) {
   }
 
   // First line gives the column names
-  fprintf(fp, "test,variable,value\n");
+  fprintf(fp, "test,variable,value,accuracy_threshold,assert_mode\n");
 
   // Iterate over all table elements
   vfc_probe_node *probe = NULL;
-  for (size_t i = 0; i < probes->map->capacity; i++) {
```
Comment (titeup) [inline src/common/vfc_probes.c:216]: why?

File: doc/06-Postprocessing.md [hunk 3]
```diff
@@ -141,10 +149,35 @@ To export your variables to a file, `vfc_probes` relies on the
...
+backend and the assert types. Backends are separated in two categories, non-
+deterministic (such as the MCA backend) and deterministic (such as the VPREC backend),
+and asserts precision can be absolute (by default) or relative. This results in four
+different conditions to validate the probe, which are summed up in the following
+table :
+
+| |Absolute|Relative|
+--- | --- | ---
+|Non-deterministic|Standard deviation < Target | Standard deviation / Empirical average < Target
+|Deterministic|IEEE value -  Backend value < Target|(IEEE value -  Backend value) / IEEE value < Target
```
Comment (pablooliveira) [inline doc/06-Postprocessing.md:167]: Please include the absolute values

File: src/common/vfc_probes.c [hunk 4]
```diff
@@ -96,14 +104,15 @@ void vfc_free_probes(vfc_probes *probes) {
 
   // Before freeing the map, iterate manually over all items to free the keys
   vfc_probe_node *probe = NULL;
-  for (size_t i = 0; i < probes->map->capacity; i++) {
```
Comment (titeup) [inline src/common/vfc_probes.c:99]: Why?

File: src/common/vfc_probes.c [hunk 5]
```diff
@@ -71,6 +75,10 @@ void validate_probe_key(char *str);
 // or a duplicate key), an error will be thrown.
 int vfc_probe(vfc_probes *probes, char *testName, char *varName, double val);
 
+// Similar to vfc_probe, but with an optional accuracy threshold.
+int vfc_probe_assert(vfc_probes *probes, char *testName, char *varName,
```
Comment (titeup) [inline src/common/vfc_probes.c:79]: naming? assert may be missleding since we do not abort

File: doc/06-Postprocessing.md [hunk 6]
```diff
@@ -121,11 +121,19 @@ Below is a list of the different functions you may use to manipulate the
...
-int vfc_probe(vfc_probes *probes, char *testName, char *varName,
-                  double val);
-
+// Add a new probe. If an issue with the key is detected (forbidden characters
+// or a duplicate key), an error will be thrown. (no assert)
+int vfc_probe(vfc_probes *probes, char *testName, char *varName, double val);
+
+// Similar to vfc_probe, but with an optional accuracy threshold (absolute
+// assert).
+int vfc_probe_assert(vfc_probes *probes, char *testName, char *varName,
```
Comment (pablooliveira) [inline doc/06-Postprocessing.md:130]: Rename `vfc_check` or `vfc_check_relative` ?

**Review by titeup — COMMENTED** [review]
Can we consider exit on failure for deterministic/(mca)sunchronous backend where we can determine it at runtime? I some complex code this could also avoid wasting costly execution.

File: doc/06-Postprocessing.md [hunk 7]
```diff
@@ -121,11 +121,19 @@ Below is a list of the different functions you may use to manipulate the
...
+// Add a new probe. If an issue with the key is detected (forbidden characters
+// or a duplicate key), an error will be thrown. (no check)
+int vfc_probe(vfc_probes *probes, char *testName, char *varName, double val);
+
+// Similar to vfc_probe, but with an optional accuracy threshold (absolute
+// check).
+int vfc_probe_check(vfc_probes *probes, char *testName, char *varName,
+                     double val, double accuracyThreshold);
+
+// Similar to vfc_probe, but with an optional accuracy threshold (absolute
```
Comment (pablooliveira) [inline doc/06-Postprocessing.md:133]: absolute -> relative

File: doc/06-Postprocessing.md [hunk 8]
```diff
@@ -141,10 +149,35 @@ To export your variables to a file, `vfc_probes` relies on the
...
+different conditions to validate the probe, which are summed up in the following
+table :
+
+| |Absolute|Relative|
+--- | --- | ---
+|Non-deterministic|\|Standard deviation\| < Target | \|Standard deviation\| / \|Empirical average\| < Target
+|Deterministic|\|IEEE value\| -  \|Backend value\| < Target|(\|IEEE value\| -  \|Backend value\|) / \|IEEE value\| < Target
+
+By default, the standard deviation is used to estimate the error, and must be
+inderior to the accuracy threshold for the probe to pass (or inferior to the
```
Comment (pablooliveira) [inline doc/06-Postprocessing.md:170]: inderior -> inferior

File: doc/06-Postprocessing.md [hunk 9]
```diff
@@ -141,10 +149,35 @@ To export your variables to a file, `vfc_probes` relies on the
...
+std. dev. / avg. quotient in the case of a relative check).
+But in the case of a deterministic backend, it is impossible to compute the standard
```
Comment (pablooliveira) [inline doc/06-Postprocessing.md:172]: "it is impossible to compute the standard deviation" -> "the standard deviation is undefined"

File: doc/06-Postprocessing.md [hunk 10]
```diff
@@ -141,10 +149,35 @@ To export your variables to a file, `vfc_probes` relies on the
...
+deviation and the average of the probe to estimate the error. Instead , the
```
Comment (pablooliveira) [inline doc/06-Postprocessing.md:173]: Instead , -> Instead,

**Review by pablooliveira — CHANGES_REQUESTED** [review]
LGTM. Please fix some typos in the documentation and this is good to merge.

**Review by pablooliveira — APPROVED** [review]
LGTM :tada:

**Pull request merged**